crates/pallet-domains/src/block_tree.rs

```diff
@@ -380,6 +393,51 @@ pub(crate) fn process_execution_receipt<T: Config>(
     Ok(None)
 }
 
+type TransferTrackerError<T> =
+    <<T as Config>::DomainsTransfersTracker as DomainsTransfersTracker<BalanceOf<T>>>::Error;
+
+/// Updates domain transfers for following scenarios
+/// 1. Block fees are burned on domain
+/// 2. Confirming incoming XDM transfers to the Domain
+/// 3. Noting outgoing transfers from the domain
+/// 4. Cancelling outgoing transfers from the domain.
+fn update_domain_transfers<T: Config>(
+    domain_id: DomainId,
+    transfers: &Transfers<BalanceOf<T>>,
+    block_fees: BalanceOf<T>,
+) -> Result<(), TransferTrackerError<T>> {
+    let Transfers {
+        transfers_in,
+        transfers_out,
+        transfers_reverted,
+    } = transfers;
+
+    // confirm incoming transfers
+    let er_chain_id = ChainId::Domain(domain_id);
+    transfers_in
+        .iter()
+        .try_for_each(|(from_chain_id, amount)| {
+            T::DomainsTransfersTracker::confirm_transfer(*from_chain_id, er_chain_id, *amount)
+        })?;
+
+    // note outgoing transfers
+    transfers_out.iter().try_for_each(|(to_chain_id, amount)| {
+        T::DomainsTransfersTracker::note_transfer(er_chain_id, *to_chain_id, *amount)
+    })?;
+
+    // cancel existing transfers
+    transfers_reverted
+        .iter()
+        .try_for_each(|(to_chain_id, amount)| {
+            T::DomainsTransfersTracker::cancel_transfer(er_chain_id, *to_chain_id, *amount)
```

Comment:
A potential issue here, it seems the canceling of an XDM transfer is controlled by the src domain.

Suppose there is a compromised domain A, the malicious operator first produces a valid ER that contains a valid XDM that transfers from domain A to domain B and a second fraudulent ER that claims that XDM has failed thus canceling the transfer. Both ERs are confirmed on the consensus chain and then the valid ER from domain B that claims the XDM is successful will be rejected since the XDM transfer is already cancelled by the fraudulent ER from domain A. As a result, domain B is attacked by domain A and stuck forever.

The key point here seems to be to define a way for the src and dest domains to reach consensus about the XDM failure 🤔.

Reply:
Interesting scenario. Will adjust it accordngly